src/github.com/getlantern/flashlight/client/chained.go

```diff
@@ -103,6 +105,79 @@ func (s *ChainedServerInfo) Dialer(deviceID string) (*balancer.Dialer, error) {
 
 			return conn, nil
 		},
+		Check: func() bool {
+			return s.check(d, deviceID)
+		},
 		OnRequest: ccfg.OnRequest,
 	}, nil
 }
+
+func (s *ChainedServerInfo) attachHeaders(req *http.Request, deviceID string) {
+	authToken := s.AuthToken
+	if ForceAuthToken != "" {
+		authToken = ForceAuthToken
+	}
+	if authToken != "" {
+		req.Header.Add("X-LANTERN-AUTH-TOKEN", authToken)
+	}
+	req.Header.Set("X-LANTERN-DEVICE-ID", deviceID)
+}
+
+func (s *ChainedServerInfo) updateCheckTarget(addr string) {
+	host, port, e := net.SplitHostPort(addr)
+	if e != nil {
+		log.Errorf("failed to split port from %s", addr)
+		return
+	}
+	if strings.HasSuffix(host, internalSiteSuffix) || port != "80" {
+		log.Tracef("Skip setting %s as check target", addr)
+		return
+	}
+	s.checkTarget.Store(addr)
+}
+
+func (s *ChainedServerInfo) check(dial func(string, string) (net.Conn, error), deviceID string) bool {
+	rt := &http.Transport{
+		DisableKeepAlives: true,
+		Dial:              dial,
+	}
+	var url string
+	checkTarget, isSet := s.checkTarget.Load().(string)
+	if !isSet {
+		url = "http://ping-chained-server"
+	} else {
+		url = fmt.Sprintf("http://%s/index.html", checkTarget)
+	}
+	req, err := http.NewRequest("HEAD", url, nil)
+	if err != nil {
+		log.Errorf("Could not create HTTP request: %v", err)
+		return false
+	}
+	if !isSet {
+		req.Header.Set("X-Lantern-Ping", "small")
+	}
+
+	s.attachHeaders(req, deviceID)
+	ok, timedOut, _ := withtimeout.Do(60*time.Second, func() (interface{}, error) {
+		resp, err := rt.RoundTrip(req)
+		if err != nil {
+			log.Debugf("Error testing dialer %s to %s: %s", s.Addr, url, err)
+			return false, nil
+		}
+		if err := resp.Body.Close(); err != nil {
+			log.Debugf("Unable to close response body: %v", err)
+		}
+		good := resp.StatusCode < 500
```

Comment:
This is a little unusual, but I guess the point is that you're just looking for any response from the origin.  Might be good to put a comment here about it.